flow/include/flow/OwningResource.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * OwningResource.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FLOW_OWNING_REOSURCE_H
+#define FLOW_OWNING_REOSURCE_H
+
+#include "flow/FastRef.h"
+
+// Consider the following situation:
+//
+//   1. An ACTOR A0 allocates an object O
+//   2. A0 spawns another ACTOR A1, which depends on O
+//   3. A0 triggers A1 and then terminates, destroying O
+//   4. Since A1 is triggered by A0 while not knowing A0 is terminated and O is released, it would cause a SEGV error.
+//
+// In this header file, two classes
+//
+//   * ResourceOwningRef
+//   * ResourceWeakRef
+//
+// are provided. The ResourceOwningRef is the reference that "holds" the resource, When it is destructed, the resource
+// is also released; while ResourceWeakRef is the reference that "weakly holds" the resource. Before each access, it is
+// the user's responsibility to verify if the resource is still available, via the available() method.
+//
+// With the two classes, the issue above can be solved by:
+//
+//    1. A0 allocates the object O via ResourceOwningRef
+//    2. A0 forwards O to A1, via ResourceWeakRef
+//    3. Every time A1 accesses O, it will verify if the resource is still available.
+//    4. When A0 terminates, O is released and all ResourceWeakRef available() call will report the resource is not
+//       available anymore, preventing the SEGV error being raised.
+
+namespace details {
+
+// The class holding the pointer to the resource.
+// SOMEDAY: Think using std::unique_ptr
+template <typename T>
+struct Resource : public ReferenceCounted<Resource<T>>, NonCopyable {
```

Comment:
Why is this extra class necessary? Can't `ResourceRef<T>` have a `Reference<T>` field?

Reply:
If I use `Reference<T>`, then the `Reference` to the resource will not be a "weak reference", instead it would increase the reference counter to the underlying `ptr` object.

Reply:
`Reference` has a `clear` method that would be analogous to calling `reset(nullptr)` on `Resource`.

Reply:
But the `clear` method is not guaranteeing the `delete` of the resource, rather it would do `delref`. Whereas here I would like to implementing something that enforces the `unique_ptr` behavior.

Reply:
I suppose the bigger issue is that the object `T` may not be of the `ReferenceCounted` type.

For something that is, we wouldn't be exposing the `Reference` and the only way to increment the reference count would be to call `addRef` externally. If someone were to do that, then it would actually probably be safer if we were using a `Reference` here, since otherwise we could end up with two entities trying to destroy it. Either way it would be disallowed, and we could enforce that via assert. But if we want to include non-ReferenceCounted types, this seems like a moot point.